vars/githubPublishRelease.groovy

```diff
@@ -0,0 +1,137 @@
+    def result = [:]
+    try {
+        def response = httpRequest "${config.githubApiUrl}/repos/${config.githubOrg}/${config.githubRepo}/releases/latest?access_token=${TOKEN}"
+        result = readJSON text: response.content
+    } catch (e) {
+        echo "[${STEP_NAME}] This is the first release - no previous releases available ${e.getMessage()}"
+        config.addDeltaToLastRelease = false
+        result = [
+            published_at: '2017-06-21T15:11:57Z',
+            tag_name: ''
+        ]
+    }
+    return result
+}
+
+String addClosedIssue(config, TOKEN, publishedAt){
+    if (config.customFilterExtension) {
+        config.customFilterExtension = "&${config.customFilterExtension}"
+    }
+
+    def response = httpRequest "${config.githubApiUrl}/repos/${config.githubOrg}/${config.githubRepo}/issues?access_token=${TOKEN}&per_page=100&state=closed&direction=asc&since=${publishedAt}${config.customFilterExtension}"
+    def result = ''
+
+    content = readJSON text: response.content
+
+    //list closed pull-requests
+    result += '<br />**List of closed pull-requests since last release**<br />'
+    for (def item : content) {
+        if (item.pull_request && !isExcluded(item, config.excludeLabels)) {
+            result += "[#${item.number}](${item.html_url}): ${item.title}<br />"
+        }
+    }
+    //list closed issues
+    result += '<br />**List of closed issues since last release**<br />'
+    for (def item : content) {
+        if (!item.pull_request && !isExcluded(item, config.excludeLabels)) {
+            result += "[#${item.number}](${item.html_url}): ${item.title}<br />"
+        }
+    }
+    return result
+}
+
+String addDeltaToLastRelease(config, latestTag){
+    def result = ''
+    //add delta link to previous release
+    result += '<br />**Changes**<br />'
+    result += "[${latestTag}...${config.version}](${config.githubServerUrl}/${config.githubOrg}/${config.githubRepo}/compare/${latestTag}...${config.version}) <br />"
+    return result
+}
+
+void postNewRelease(config, TOKEN, releaseBody){
+    releaseBody = releaseBody.replace('"', '\\"')
+    //write release information
+    def data = "{\"tag_name\": \"${config.version}\",\"target_commitish\": \"master\",\"name\": \"${config.version}\",\"body\": \"${releaseBody}\",\"draft\": false,\"prerelease\": false}"
+    try {
+        httpRequest httpMode: 'POST', requestBody: data, url: "${config.githubApiUrl}/repos/${config.githubOrg}/${config.githubRepo}/releases?access_token=${TOKEN}"
+    } catch (e) {
+        echo """[${STEP_NAME}] Error occured when writing release information
+---------------------------------------------------------------------
+Request body was:
+---------------------------------------------------------------------
+${data}
+---------------------------------------------------------------------"""
+        throw e
+    }
+}
+
+boolean isExcluded(item, excludeLabels){
+    def result = false
+    if (!excludeLabels.isEmpty()) {
```

Comment:
[nit picking] ifEmptyCheck not strongly required. In case we have an empty iterable the each closure is simply not executed.